server/datastore/mysql/software.go

```diff
@@ -76,16 +80,152 @@ func (d *Datastore) SaveHostSoftware(host *fleet.Host) error {
 	return nil
 }
 
-func (d *Datastore) LoadHostSoftware(host *fleet.Host) error {
-	host.HostSoftware = fleet.HostSoftware{Modified: false}
+func nothingChanged(current []fleet.Software, incoming []fleet.Software) bool {
+	if len(current) != len(incoming) {
+		return false
+	}
+
+	currentBitmap := make(map[string]bool)
+	for _, s := range current {
+		currentBitmap[softwareToUniqueString(s)] = false
+	}
+	for _, s := range incoming {
+		if _, ok := currentBitmap[softwareToUniqueString(s)]; !ok {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (d *Datastore) applyChangesForNewSoftware(tx *sqlx.Tx, host *fleet.Host) error {
+	storedCurrentSoftware, err := d.hostSoftwareFromHostID(tx, host.ID)
+	if err != nil {
+		return errors.Wrap(err, "loading current software for host")
+	}
+
+	if nothingChanged(storedCurrentSoftware, host.Software) {
+		return nil
+	}
+
+	current := softwareSliceToIdMap(storedCurrentSoftware)
+	incoming := softwareSliceToSet(host.Software)
+
+	if err = d.deleteUninstalledHostSoftware(tx, host.ID, current, incoming); err != nil {
+		return err
+	}
+
+	if err = d.insertNewInstalledHostSoftware(tx, host.ID, current, incoming); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *Datastore) deleteUninstalledHostSoftware(
+	tx *sqlx.Tx,
+	hostID uint,
+	currentIdmap map[string]uint,
+	incomingBitmap map[string]bool,
+) error {
+	var deletesHostSoftware []interface{}
+	deletesHostSoftware = append(deletesHostSoftware, hostID)
+
+	for currentKey := range currentIdmap {
+		if _, ok := incomingBitmap[currentKey]; !ok {
+			deletesHostSoftware = append(deletesHostSoftware, currentIdmap[currentKey])
+			// TODO: delete from software if no host has it
+		}
+	}
+	if len(deletesHostSoftware) > 1 {
+		sql := fmt.Sprintf(
+			`DELETE FROM host_software WHERE host_id = ? AND software_id IN (%s)`,
+			strings.TrimSuffix(strings.Repeat("?,", len(deletesHostSoftware)-1), ","),
+		)
+		if _, err := tx.Exec(sql, deletesHostSoftware...); err != nil {
+			return errors.Wrap(err, "delete host software")
+		}
+	}
+	return nil
```

Comment:
What do you think of this clarification?
```suggestion
	var deletesHostSoftware []uint

	for currentKey := range currentIdmap {
		if _, ok := incomingBitmap[currentKey]; !ok {
			deletesHostSoftware = append(deletesHostSoftware, currentIdmap[currentKey])
			// TODO: delete from software if no host has it
		}
	}
	if len(deletesHostSoftware) == 0 {
		return nil
	}
	sql := fmt.Sprintf(
		`DELETE FROM host_software WHERE host_id = ? AND software_id IN (%s)`,
		strings.TrimSuffix(strings.Repeat("?,", len(deletesHostSoftware)), ","),
	)
	if _, err := tx.Exec(sql, hostID, deletesHostSoftware...); err != nil {
		return errors.Wrap(err, "delete host software")
	}
	return nil
```

Reply:
+1 to the early exit of `if len(deletesHostSoftware) == 0 {`

The rest doesn't seem to work, the deletesHostSoftware needs to be of interface{} because otherwise it needs a cast for each element. And `tx.Exec(sql, hostID, deletesHostSoftware...)` doesn't seem to work either sadly.

The fun fact here is that I did try those things, so we are very much aligned :)